include/lang_utils/tuple.h

```diff
@@ -2,224 +2,135 @@
 
 template <typename ...TUPLES>
-struct tuple_sizes_equal {
-    template <typename FIRST, typename ...REMAINING>
-    struct helper {
-        static const size_t value =
-            std::tuple_size<typename std::decay<FIRST>::type>::value;
-        static_assert(tuple_sizes_equal<REMAINING...>::value == value,
-                      "tuple sizes not equal");
-    };
-
-    static const size_t value = helper<TUPLES...>::value;
-};
-
-template <typename FUNC, typename ...TUPLES>
-class foreach_tuple_helper {
-    template <typename F, typename ...T>
-    friend void foreach_tuple(F&&, T&&...);
-
-    void operator()(std::index_sequence<>, FUNC &&func,
-                    TUPLES&& ...tuples) { }
-
-    template <size_t POS, size_t ...INDICES>
-    void operator()(std::index_sequence<POS, INDICES...>, FUNC &&func,
-                  TUPLES&& ...tuples) {
-        func(std::get<POS>(tuples)...);
-        (*this)(std::index_sequence<INDICES...>(),
-                std::forward<FUNC>(func),
-                std::forward<TUPLES>(tuples)...);
-    }
-};
-
-template <typename FUNC, typename ...TUPLES>
-void foreach_tuple(FUNC &&func, TUPLES&& ...tuples) {
-    static const size_t size = tuple_sizes_equal<TUPLES...>::value;
+struct tuple_sizes_equal : public
+    std::integral_constant<size_t, tuple_sizes_equal_v<TUPLES...>> {};
 
-    foreach_tuple_helper<FUNC, TUPLES...>()(std::make_index_sequence<size>(),
-                                            std::forward<FUNC>(func),
-                                            std::forward<TUPLES>(tuples)...);
+template <size_t N, typename... TUPLES>
+auto slice_tuples(TUPLES&&... tuples) {
+    return std::tie(std::get<N>(std::forward<TUPLES>(tuples))...);
 }
 
-template <typename FUNC, typename ...TUPLES>
-class foreach_tuple_i_helper {
-    template <typename F, typename ...T>
-    friend void foreach_tuple_i(F&&, T&&...);
-    
-    void operator()(std::index_sequence<>, FUNC &&func,
-                    TUPLES&& ...tuples) { }
-
-    template <size_t POS, size_t ...INDICES>
-    void operator()(std::index_sequence<POS, INDICES...>, FUNC &&func,
-                  TUPLES&& ...tuples) {
-        func(POS, std::get<POS>(tuples)...);
-        (*this)(std::index_sequence<INDICES...>(),
-                std::forward<FUNC>(func),
-                std::forward<TUPLES>(tuples)...);
-    }
-    
+enum class pos_kind {
+    none,
+    compile_time,
+    run_time
 };
 
-template <typename FUNC, typename ...TUPLES>
-void foreach_tuple_i(FUNC &&func, TUPLES&& ...tuples) {
-    static const size_t size = tuple_sizes_equal<TUPLES...>::value;
+template <pos_kind PK, typename FUNC>
+auto adapt_impl(FUNC&& func) {
+    if constexpr (PK == pos_kind::run_time) {
+        return [&func]<size_t i>() {
+            return [&func]<typename... ARGS>(ARGS&&... args) -> decltype(auto) {
+                return std::forward<FUNC>(func)(i, std::forward<ARGS>(args)...);
+            };
+        };
```

Comment:
I'm using a feature that will be in C++20 but is supported today by GCC (but not clang): template syntax for lambdas.

Reply:
I just noticed the adapt_impl thing with the constexpr on an enum. You really like making compile time logic look like run time logic. What about using partial specialization for this? I feel like it looks more explicit:
```
template <pos_kind PK>
struct adapt_impl;

template <>
struct adapt_impl<pos_kind::run_time> {
    template <typename FUNC>
    auto operator()(FUNC&& func) {
        return [&func]<size_t i>() {
            return [&func]<typename... ARGS>(ARGS&&... args) -> decltype(auto) {
                return std::forward<FUNC>(func)(i, std::forward<ARGS>(args)...);
            };
        };
    }
};

template <>
struct adapt_impl<pos_kind::compile_time> {
    template <typename FUNC>
    auto operator()(FUNC&& func) {
        return [&func]<size_t i>() {
            return [&func]<typename... ARGS>(ARGS&&... args) -> decltype(auto) {
                return std::forward<FUNC>(func).template operator()<i>(std::forward<ARGS>(args)...);
            };
        };
    }
};

template <>
struct adapt_impl<pos_kind::none> {
    template <typename FUNC>
    auto operator()(FUNC&& func) {
        return [&func]<size_t>() {
            return [&func]<typename... ARGS>(ARGS&&... args) -> decltype(auto) {
                return std::forward<FUNC>(func)(std::forward<ARGS>(args)...);
            };
        };
    }
};
```